File: project/.sdk/tm/rb/test/runner.rb
# ProjectName SDK test runner

require 'json'

module ProjectNameTestRunner
  @env = {}

  def self.load_env_local
    env_file = File.join(File.dirname(__FILE__), '..', '..', '.env.local')
    return unless File.exist?(env_file)

    File.readlines(env_file).each do |line|
      line = line.strip
      next if line.empty? || line.start_with?('#')
      key, val = line.split('=', 2)
      next unless key && val
      @env[key.strip] = val.strip
    end
  end

  def self.getenv(key)
    @env[key] || ENV[key]
  end

  def self.env_override(m)
    live = getenv("PROJECTNAME_TEST_LIVE")
    override = getenv("PROJECTNAME_TEST_OVERRIDE")

    if live == "TRUE" || override == "TRUE"
      m.each_key do |key|
        envval = getenv(key)
        if envval && !envval.empty?
          envval = envval.strip
          if envval.start_with?('{')
            begin
              parsed = JSON.parse(envval)
              m[key] = parsed
              next
            rescue JSON::ParserError
            end
          end
          m[key] = envval
        end
      end
    end

    explain = getenv("PROJECTNAME_TEST_EXPLAIN")
    m["PROJECTNAME_TEST_EXPLAIN"] = explain if explain && !explain.empty?

    m
  end

  def self.entity_list_to_data(list)
    out = []
    list.each do |item|
      if item.is_a?(Hash)
        out << item
      elsif item.respond_to?(:data_get)
        d = item.data_get
        out << d if d.is_a?(Hash)
      end
    end
    out
  end
end

File: project/.sdk/tm/py/test/runner.py
# ProjectName SDK test runner

from __future__ import annotations
import os
import json

from utility.voxgig_struct import voxgig_struct as vs


class ProjectNameTestRunner:
    _env = {}

    @staticmethod
    def load_env_local():
        try:
            with open("../../.env.local", "r") as f:
                content = f.read()
        except (FileNotFoundError, IOError):
            return

        for line in content.splitlines():
            line = line.strip()
            if line == "" or line.startswith("#"):
                continue
            eq_idx = line.find("=")
            if eq_idx < 0:
                continue
            key = line[:eq_idx].strip()
            val = line[eq_idx + 1:].strip()
            ProjectNameTestRunner._env[key] = val

    @staticmethod
    def getenv(key):
        val = ProjectNameTestRunner._env.get(key)
        if val is not None:
            return val
        return os.environ.get(key)

    @staticmethod
    def env_override(m):
        live = ProjectNameTestRunner.getenv("PROJECTNAME_TEST_LIVE")
        override = ProjectNameTestRunner.getenv("PROJECTNAME_TEST_OVERRIDE")

        if live == "TRUE" or override == "TRUE":
            for key in list(m.keys()):
                envval = ProjectNameTestRunner.getenv(key)
                if envval is not None and envval != "":
                    envval = envval.strip()
                    if envval.startswith("{"):
                        try:
                            parsed = json.loads(envval)
                            if parsed is not None:
                                m[key] = parsed
                                continue
                        except Exception:
                            pass
                    m[key] = envval

        explain = ProjectNameTestRunner.getenv("PROJECTNAME_TEST_EXPLAIN")
        if explain is not None and explain != "":
            m["PROJECTNAME_TEST_EXPLAIN"] = explain

        return m

    @staticmethod
    def entity_list_to_data(lst):
        out = []
        for item in lst:
            if isinstance(item, dict):
                out.append(item)
            elif hasattr(item, "data_get") and callable(item.data_get):
                d = item.data_get()
                if isinstance(d, dict):
                    out.append(d)
            else:
                out.append(item)
        return out
